### DISCUSSION TITLE
What would have been the easiest way to realize I have to pass Query params to a handler?

### DISCUSSION THREAD

**Discussion by TeemuKoivisto** [tokio-rs/axum#2052] [Q&A] [answered]
So I have a handler like

```rs
pub async fn ws_handler(
    ws: WebSocketUpgrade,
    Query(params): Query<Params>,
    State(state): State<Arc<Context>>,
) -> AxumResponse {
  ...
}
```

and I spend a good while debugging why it didn't work as I copy-pasted it from a project to another. Until I figured out that I wasn't passing the JWT as I did originally `ws://localhost:1234?jwt=abc`, thus causing the handler to always immediately return 400 and not even allowing me to print stuff in the handler.

I added some tracing with tower which just showed a 400 response. Was there an easy way to realize this before I wasted time fiddling about? Or is this just one of those things that you just should know beforehand? To me it wasn't obvious but I guess tower showing a response of 400 should have rung some bells.

Oh well, hopefully this helps someone else in the future.

**Comment by davidpdrsn** [comment, marked as answer]
The response body would have told you. You can also enable tracing to built in extractors like documented here https://docs.rs/axum/latest/axum/extract/index.html#logging-rejections

**Reply by TeemuKoivisto** [reply]
Okay. Yeah figured there ought to have been some kind of error message. Since it was a binary websocket handler though, couldn't see anything from the client-side.

That tracing would have been Godsent though! Thanks for sharing, I'll keep it mind next time 👍.
